### PR TITLE
Block Supports: Refactor styles rendering to utility function and add to REST API responses

### PR DESCRIPTION
**Pull request by andrewserong**
## Description

Fixes #35413 and possible fix for https://github.com/WordPress/gutenberg/issues/35331

This PR explores an idea (and alternative to) #35413 to render block supports styles via a common utility function. This function adds the styles via the `wp_footer` action, and also registers the styles with a filter that is then used to add the styles to API responses for public post types.

### Background

Currently the Layout, Duotone, and Elements block supports each feature some form of rendering where either styles or `svg` markup are appended to the rendered post/page via `wp_footer`. For these styles to be available within a server rendered block in the editor, they also need to be returned in REST API requests.

The goal in this exploration is to see whether we can come up with a common function to render out these styles / markup from block supports, and to also support rendering in a REST API request. The current approach to achieve the latter is by registering an additional REST API field for public post types, `block_supports_styles`:

image

The end result is that we fix the bug reported in #35376 where block supports styles are not rendered in the post content block within the site editor, when it is a child of the Query Loop block. 

## Screenshots 

The following screenshots are from the Site Editor, viewing the Post Content block within the Query Loop block. The rendered post content contains:

* A paragraph block with a custom link color
* An image block with Duotone filter applied
* Social Icons block

| Before | After |
| --- | --- |
| image | image |

#### Before applying this PR:

1. Activate a blocks-based theme, e.g. TT1-Blocks
2. Add a post and insert a Social Icons block, and add a bunch of icons with links
3. Add an image block and apply the Duotone filter
4. Add a paragraph with a link, and set a custom link color
3. Publish the post and open up the Site Editor
4. Assuming your theme uses the Query Loop + Post Content blocks in the main index template, you should see the issue reported in #35376 where the layout spacing styles used by the Social Icons block is not rendered within the editor.

#### After applying this PR:

1. Reload the site editor, and you should see that the styling issues are resolved as in the above screenshots
2. View the front end of your site — the styles should still render correctly, and if you view source, you should see that the wp-container styles (e.g. `<style>.wp-container-615e737c1a8f3 > *`) are still rendered in the footer, after script tags, etc.
3. You can also test the REST API endpoint directly by grabbing the post id of your published post. Then, in the site editor, once you've reloaded the page, search the Network tab of dev tools for your post id. You should find a network request for the `posts` API endpoint, used to render the Post Content block within the editor. E.g. on my test site, there's a URL like: http://localhost:8888/index.php?rest_route=%2Fwp%2Fv2%2Fposts%2F1105&_locale=user
4. If you inspect the JSON returned by that endpoint, you should see the block supports styles are now rendered in a `block_supports_styles` field

## To-dos

* [x] Refactor the Layout support to use the common function to render styles
* [x] Refactor the Duotone support to use the common function to render styles
* [x] Refactor the Elements support to use the common function to render styles
* [x] Update Post Content block in the editor to append styles from the API endpoint
* [ ] Add PHP tests to cover the behaviour

## Types of changes

Bug fix (non-breaking change which fixes an issue)

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] I've tested my changes with keyboard and screen readers. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

**Comment by andrewserong** [comment]
@youknowriad @spacedmonkey — just a heads-up this is where I'm up to with refactoring the block supports styles to be rendered correctly in the site editor. I've gone with Jonny's suggestion to add a separate field to the REST API responses, and then in the Post Content block, we then append those styles in read only mode.

I haven't yet added tests (but am happy to, and would like to). I'm keen for feedback on the approach before I continue polishing, whenever you have time to take a look 🙂

File: packages/block-library/src/post-content/edit.js [hunk 1]
```diff
@@ -12,29 +12,35 @@ import {
...
 		</div>
 	) : (
-		<div { ...blockProps }>
-			<RawHTML key="html">{ content?.rendered }</RawHTML>
-		</div>
+		<RawHTML { ...blockProps } key="html">
+			{ blockSupportsStyles && content?.rendered
+				? content.rendered + blockSupportsStyles
+				: content?.rendered }
+		</RawHTML>
```
Comment (ellatrix) [inline packages/block-library/src/post-content/edit.js:43]: What if layout, duotone etc. is rendered outside of the post content?

Comment (andrewserong) [reply packages/block-library/src/post-content/edit.js:43]: Good question, thanks for taking a look @ellatrix. This particular line just deals with the Post Content block, so grabbing `blockSupportsStyles` as a field from the posts endpoint currently works, but this PR does assume we're dealing with post content. Outside of the Post Content block within the site editor, the block supports still appear to be rendering okay in my testing so far, but I haven't tested with other REST API based server-rendered blocks, so there are likely still many gaps.

We could also add the field to other resources that need it, e.g. [comments](https://developer.wordpress.org/rest-api/reference/comments/), possibly the [block renderer](https://developer.wordpress.org/rest-api/reference/rendered-blocks/), too?

I've struggled a bit with coming up with a clean solution to ensure that the rendered block supports styles are always available and will be used in each potential case. In this PR, we mostly solve for the problem of rendering within the Post Content block in the editor, however I can imagine there will be a number of edge cases where we'll need to be intentional about how we render (and make available) the block supports styles. If this approach looks viable, we can always follow up with each of the edge cases as we encounter them, but I'm aware that approach could be brittle (and not necessarily great for other consumers of the REST API).

Let me know if you can think of a better way to expose the styles that we're currently rendering out to `wp_footer`, I'm very happy to rework this PR if we can come up with better ways to handle it! 🙂

**Comment by andrewserong** [comment]
Update: the approach proposed in this PR adds a field to the API responses for the Posts and Pages API endpoints. With a bit more time to think about this over the past couple of weeks, I think in general, how we ship block supports / global styles via API endpoints needs a bit more thought to come up with a generalisable solution. For example, in this PR, we really only deal with posts and pages, but comments or other post types also support blocks, so these would need to be factored in. And in the longer-term, we'd ideally want to think through carefully how developers are supposed to retrieve and combine styles with content.

So, as a shorter-term fix for the issue at hand, I've opened up a separate PR to switch the Post Content block over to render its readonly state as live blocks in the editor, instead of raw html. This seems to fix the main issue that inspired this PR (rendering block supports styles in the site editor for the post content block). I'm keen for feedback if anyone gets time to take a look: https://github.com/WordPress/gutenberg/pull/35863

If that approach looks viable (to fix the visible bug in time before 5.9) then we could potentially close out this PR, and defer the discussion surrounding returning styles via API endpoints to another time.

**Comment by TimothyBJacobs** [comment]
Related: https://core.trac.wordpress.org/ticket/48654, https://core.trac.wordpress.org/ticket/46379

> If that approach looks viable (to fix the visible bug in time before 5.9) then we could potentially close out this PR, and defer the discussion surrounding returning styles via API endpoints to another time.

This seems ideal. Getting those endpoints right is important and I think too difficult to land in a few weeks.

**Comment by andrewserong** [comment]
Thanks for the links to the prior Trac tickets, Timothy!

It looks like we're honing in on a workable approach in https://github.com/WordPress/gutenberg/pull/35863 now, so I'm going to close out this PR. Thanks everyone!

**Pull request closed without merging**